### DISCUSSION TITLE
With React.Component, how do I access MobX state in a callback

### DISCUSSION THREAD

**Discussion by heath-freenome** [mobxjs/mobx#2888] [Q&A] [answered]
I am doing something like the code below...

QUESTION: Why isn't the callback function using the `observer()` nature of the class? If I don't include the `runInAction()` then I get this error: `[mobx] Observable FooStore@2.values being read outside a reactive context`

NOTE the component below is a very simplified version of what I'm doing, so I don't want to switch to a stateless functional component
```
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { runInAction } from 'mobx';
import { observer } from 'mobx-react';

class FooStore {
  values = { foo: 'bar' };

  constructor() {
    makeAutoObservable(this);
  }
}

class Foo extends Component {
  propTypes = {
    store: PropTypes.instanceOf(FooStore).isRequired,
  }; 
  onCheck = (event) => {
    const { store } = this.props;
    const { checked } = event.target;
    let newValue = {};
    if (checked) {
      runInAction(() => {
        newValue = { ...store.values };
      });
    }
    this.update(newValue);
  }
  update(values)
    // do something with values
  };
  render() {
    const { store } = this.props;
    return (
      <form>
        <span>{JSON.stringify(store.values)}</span>
        <input type="checkbox" id="chk" name="chk" onChange={this.onCheck} />
        <label for="chk">Use the above values</label>
      </form>
    );
  }
}

export default observer(Foo);
```

**Comment by mweststrate** [comment, marked as answer]
I couldn't parse the question. I think you are asking for: 

```
onCheck = action((event) => {
   // etc
 })
```

**Reply by heath-freenome** [reply]
I guess only the `render()` function is explicitly an action?

**Reply by mweststrate** [reply]
It's not an action at all, it is a derivation. See https://mobx.js.org/the-gist-of-mobx.html for the conceptual overview. The original warning you get is probably from enabling this flag setting https://mobx.js.org/configuration.html#observablerequiresreaction-boolean, which is not the default but forces you to have a derivation or action everywhere where you are dealing with observables. Since it is in principle completely fine to _read_ observables outside an action, we don't enforce it by default, so I think that is something that is done in your project setup specifically. (It might help with catching a forgotten `observer` though, that is what the flag is for)

**Reply by heath-freenome** [reply]
@mweststrate I see... I turn that warning on in my tests to ensure we are handling MobX properly. Hmmm, so to silence these warnings, is it ok to use the `action()`? or is there another wrapper function that would be helpful?

**Comment by codeBelt** [comment]
You should update the store data via an action on the store like:

```js
export class FooStore {
  values = { foo: "bar" };

  constructor() {
    makeAutoObservable(this);
  }

  updateValue(value) {
    this.values = value;
  }
}

class Foo extends Component {
  onCheck = (event) => {
    const { checked } = event.target;

    this.props.store.updateValue({ foo: checked });
  };

  render() {
    const { store } = this.props;

    return (
      <form>
        <span>{JSON.stringify(store.values)}</span>
        <input type="checkbox" id="chk" name="chk" onChange={this.onCheck} />
        <label htmlFor="chk">Use the above values</label>
      </form>
    );
  }
}

export default observer(Foo);
```

Check out your example in codesandbox:

https://codesandbox.io/s/mobxdiscussions2888-q7bkb

**Reply by heath-freenome** [reply]
Actually, in this component, my store is read-only. I am using the `values` prop for other things

**Comment by mweststrate** [comment]
Action is fine
